File: 0146-lru-cache/0146-lru-cache.go
type LRUCache struct {
    Map map[int]*Node
    Queue LinkedList
    Cap int
}

type LinkedList struct {
    Head *Node
    Tail *Node
}

func (this *LinkedList) moveToFront(node *Node) {
    this.remove(node)
    this.add(node)
}

func (this *LinkedList) removeLast() *Node{
    evictNode := this.Tail.Prev
    this.remove(evictNode)
    return evictNode
}

func (this *LinkedList) add(node *Node){
    node.Next = this.Head.Next
    node.Prev = this.Head
    node.Next.Prev = node
    this.Head.Next = node
}

func (this *LinkedList) remove(node *Node){
    node.Prev.Next = node.Next
    node.Next.Prev = node.Prev
    node.Prev = nil
    node.Next = nil
}

type Node struct {
    Key int
    Value int
    Next *Node
    Prev *Node
} 


func Constructor(capacity int) LRUCache {
    head := &Node{
        Key: -1,
    }
    tail := &Node{
        Key: -1,
    }
    head.Next = tail
    tail.Prev = head
    queue := LinkedList{head, tail}
    return LRUCache{make(map[int]*Node), queue, capacity}
}


func (this *LRUCache) Get(key int) int {
    val, ok := this.Map[key]; 
    if !ok{
        return -1
    }
    this.Queue.moveToFront(val)
    return val.Value
}


func (this *LRUCache) Put(key int, value int)  {
    val, ok := this.Map[key]; 
    if ok{
        val.Value = value
        this.Queue.remove(val)
    }else{
        val = &Node{key, value, nil, nil}
        if this.Cap == len(this.Map){
            evict := this.Queue.removeLast()
            delete(this.Map, evict.Key)
        }
        this.Map[key] = val
    }
    this.Queue.add(val)
}


/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */

File: 0146-lru-cache/0146-lru-cache.java
class LRUCache {
    HashMap<Integer, Node> map;
    DoublyLinkedList list;
    int cap;
    public LRUCache(int capacity) {
        map = new HashMap<>();
        list = new DoublyLinkedList();
        cap = capacity;
    }
    
    public int get(int key) {
        Node n = map.getOrDefault(key, null);
        if(n == null){
            return -1;
        }
        list.remove(n);
        list.push(n);
        return n.val;
    }
    
    public void put(int key, int value) {
        if(map.containsKey(key)){
            Node n = map.get(key);
            n.val = value;
            list.remove(n);
            list.push(n);
            return;
        }
        if(map.size() == cap){
            Node toremove = list.pop();
            map.remove(toremove.key);
        }
        Node n = new Node(key, value);
        map.put(key, n);
        list.push(n);
    }
}

class Node{
    int key;
    int val;
    Node next;
    Node prev;
    public Node(){}
    public Node(int key, int val){
        this.key = key;
        this.val = val;
    }
}

class DoublyLinkedList{
    Node head;
    Node tail;
    int size;
    public DoublyLinkedList(){
        size = 0;
        head = new Node();
        tail = new Node();
        head.next = tail;
        tail.prev = head;
    }
    public Node pop(){
        Node oldTail = tail.prev;
        remove(oldTail);
        return oldTail;
    }

    public void remove(Node n){
        n.prev.next = n.next;
        n.next.prev = n.prev;
        n.next = null;
        n.prev = null;
        size--;
    }

    public void push(Node n){
        n.next = head.next;
        n.prev = head;
        head.next.prev = n;
        head.next = n;
        size++;
    }
}



/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */